docs/internal/GeoffsRecoilAssessment.md

```diff
@@ -0,0 +1,176 @@
+This document describes my assessment of moving to [Recoil](http://recoiljs.org) for application state management based on converting an application to Recoil.  It provides the pros and cons I discovered, what I think influences the cost of adoption/conversion, my best practices, and an example of implementing the Dispatcher pattern.
+
+# Background : Assessment Process
+I have a React application (a text labeler) initially written using MobX. The application state followed the typical approach for MobX - a hierarchy of models which use @observable and @computed. Update to model state was done with @action methods in the models (although following a dispatch/handler pattern would have been an easy alternative). The React components for the app were a mixture of container components, presentation components, and pure components. The container components were wrapped with the MobX observer() method to respond to state changes.  The presentation components relied on prop changes and React.useEffect.  The pure components relied only on prop changes.
+
+I rewrote the application without MobX and using only using React's context API and useEffect to respond to state changes. The application state remained in a hierarchy, but without any observer/observable pattern.  The state was passed down the component hierarchy via context.  Container components detected state changes via additional useEffect calls. In order to coordinate larger reset of the application state (such as clearing most data), an application pub/sub mechanism was used to communicate between components to orchestrate the sequence of clearing state. This was because the useEffect order was difficult to guarantee.
+
+Finally, I rewrote the application to use Recoil. I flattened the application state into atoms. Because relationships between the data create a logical hierarchy, I wrote utility functions to make cross referencing between data islands easy. I then wrote selectors to shape the data for my React components. The container components leveraged useRecoilValue and useRecoilState to respond to state changes. Because the same application state change be updated from different React components, I implemented a dispatcher pattern based on useRecoilCallback.  I had several missteps as I learned how Recoil works, and discovered how my application was affected by flattening the state.
+
+# Recoil API: Pros, Cons, Unknowns, and Hazards
+
+## Overall
+- PRO: Built for hooks from the start and follows React concepts
+- PRO: Atom and selectors allow for composing applications from small pieces without introducing unnecessary dependencies.
+- PRO: Significantly reduces the complexity and code volume in components.
+- PRO: Looking toward the future of asynchronous, multi-threaded Javascript.
+- PRO: Recoil API surface is tiny
+- PRO: Lightning fast (at least for my application size)
+- CON: Still experimental, few releases, limited contributes - not yet mature or fully tested.
```

Comment:
This for me is the big reason not to jump fully into Recoil yet. As cool as it is, I'm a little wary of future API changes that would force us to go through and rewrite everything.

Reply:
Agree, and is it the only CON in overall? sounds too good to be true